File: html/src/app/animpiece.ts
import {
  COLUMNS,
  ROWS,
  SQUARE_W,
  SQUARE_H,
  FRAME_DELAY,
  JITTER,
  WTB,
  BTW,
  WHITE,
  BLACK,
  FRAMES,
  WTB_X_OFFSET,
  WTB_Y_OFFSET,
  BTW_X_OFFSET,
  BTW_Y_OFFSET,
} from './constants';
import type { Piece } from './piece';

export class AnimPiece {
  position = ROWS * COLUMNS;
  currentFrame = 0;
  direction = BTW;
  playing = false;
  piece: Piece | null = null;
  frameAt = 0;
  startAt = 0;

  constructor(direction: number) {
    this.direction = direction;
  }

  setPosition(pos: number): boolean {
    if (pos < 64) {
      this.position = pos;
      return true;
    }
    return false;
  }

  getResult(): number {
    return this.direction === BTW ? WHITE : BLACK;
  }

  animPlay(piece: Piece, seq = 0, jitter = false): void {
    this.playing = true;
    this.piece = piece;
    this.setPosition(piece.getPosition());
    piece.setColor(this.getResult());

    let startDelay = 0;
    if (jitter && seq > 0) {
      startDelay = Math.floor(Math.random() * JITTER * FRAME_DELAY);
    } else {
      startDelay = Math.floor(FRAME_DELAY * seq);
    }

    this.currentFrame = 0;
    this.startAt = performance.now() + startDelay;
    this.frameAt = this.startAt;
  }

  stopPlaying(): void {
    this.playing = false;
    this.currentFrame = 0;
    this.piece = null;
  }

  advanceFrame(now: number): boolean {
    if (!this.playing || now < this.startAt) {
      return false;
    }

    if (now >= this.frameAt + FRAME_DELAY) {
      if (this.currentFrame < FRAMES) {
        this.currentFrame += 1;
        this.frameAt = now;
        return true;
      }
      if (this.piece) {
        this.piece.setColor(this.getResult());
      }
      this.stopPlaying();
    }
    return false;
  }

  getPixel(): { x: number; y: number } {
    const x = this.position % COLUMNS;
    const y = Math.floor(this.position / ROWS);
    const xOffset = this.direction === WTB ? WTB_X_OFFSET : BTW_X_OFFSET;
    const yOffset = this.direction === WTB ? WTB_Y_OFFSET : BTW_Y_OFFSET;
    return { x: x * SQUARE_W + xOffset, y: y * SQUARE_H + yOffset };
  }
}

File: animpiece.cpp
/***************************************************************************
                          animpiece.cpp  -  description
                             -------------------
    begin                : Sat May 10 2003
    copyright            : (C) 2003 by Ilmari Heikkinen
    email                : irheikki@cc.hut.fi
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "animpiece.h"

#include <qtimer.h>
#include <qobject.h>
#include <qdatetime.h>
#include <stdlib.h>
#include <time.h>

// connect timers to animation start and frame advance, set direction
AnimPiece::AnimPiece( QCanvasPixmapArray *parray, QCanvas *canvas, unsigned int d ) :
 QObject(), Piece(parray,canvas), direction(d), playing(FALSE), STimer(new QTimer), Timer(new QTimer)
{
    // set position out of range to make position updating mandatory on first time
    position = ROWS*COLUMNS;
    // seed rand() with processor ticks
    srand( (int)clock() );
    
    connect( STimer, SIGNAL(timeout()), SLOT(startAnim()) );
    connect( Timer, SIGNAL(timeout()), SLOT(advanceFrame()) );
}

AnimPiece::~AnimPiece()
{
  STimer->stop();
  Timer->stop();
  delete STimer;
  delete Timer;
}

// move to new position. z-index according to color
bool AnimPiece::setPosition(unsigned int pos)
{
  if (pos < (64)) { /*COLUMNS*ROWS*/
      if (position != pos) {
        position = pos;
        double x_offset, y_offset;
        if (direction == WTB) {
            x_offset = WTB_X_OFFSET;
            y_offset = WTB_Y_OFFSET;
          } else {
            x_offset = BTW_X_OFFSET;
            y_offset = BTW_Y_OFFSET;
          }

          // move piece on board, x * square_width, y * square_height. plus offsets
          double x = (double)(position % COLUMNS);
          double y = (double)(position / ROWS);
          move((x*SQUARE_W)+(x_offset), (y*SQUARE_H)+(y_offset));
      }      
      int z;
      z = 100 + position;
      setZ( z );
      return true;
  }
  return false;
}

// set direction. unused.
void AnimPiece::setDirection(unsigned int d)
{
 direction = d;
}

// identifier for canvasitems
int AnimPiece::rtti() const {
  return ANIM_RTTI;
}

// returns true if the animation is playing
bool AnimPiece::isPlaying()
{
  return playing;
}

// returns the result color of the flip animation
unsigned int AnimPiece::getResult()
{
  if (direction == BTW) {
    return WHITE;
  } else {
    return BLACK;
  }
}

// set next frame unless last, in which case hide animation
// and set the piece to right color
void AnimPiece::advanceFrame()
{
  if (currentFrame < FRAMES) {
    if (currentFrame == 0) {
      show();
    }
    setFrame(currentFrame);
    currentFrame++;
  } else {
    piece->setFrame( getResult() );
    stopPlaying();
  }  
}

// stop frame timer, rewind, show resulting piece frame, hide anim
void AnimPiece::stopPlaying()
{
    playing = FALSE;
    currentFrame = 0;

    Timer->stop();
    setFrame(currentFrame);
    hide();
}

// stop animstart timer, reset animation, set piece, start playing
void AnimPiece::animPlay(Piece* p, double seq, bool jitter)
{
  STimer->stop();
  Timer->stop();
  playing = TRUE;  
  piece = p;
  setPosition(p->getPosition());
  piece->setColor( getResult() );
  // anim start time jitter for ze bombardment.
  int startDelay;
  if (jitter && seq > 0) { // fire off first in sequence instantly
    startDelay = (int)(((float)rand()/(float)0x7fffffff)*JITTER*FRAME_DELAY);
  } else {
    startDelay = (int)(FRAME_DELAY * seq);
  }
  STimer->start( startDelay, TRUE );
}

void AnimPiece::startAnim() {
  Timer->start( FRAME_DELAY, FALSE );
  advanceFrame();
}
